Changes-for-displaying-proc-files-related-output

vedansh-bhardwaj opened the pull request:
This Commit has Changes related to display Proc files output.

Please review and let us know if any changes required.
Thanks

vedansh-bhardwaj committed:
Required changes for displaying proc related output

This patch updates the logic used to display information through the
proc files in ksmbd. The goal is to make the output more readable,
structured, and aligned with kernel formatting conventions.

Key changes include:
- Refactored code handling proc file reads for better maintainability.
- Updated data structures to ensure proper display of stats and status.
- Improved consistency of field names and value formatting.

Signed-off-by: Bahubali B Gumaji  <bahubali.bg@samsung.com>
Signed-off-by: Sang-Soo Lee  <constant.lee@samsung.com>

mmakassikis commented on Makefile:
```diff
@@ -26,13 +26,18 @@ ksmbd-y :=	unicode.o auth.o vfs.o vfs_cache.o connection.o crypto_ctx.o \
+ccflags-y += -Wno-implicit-function-declaration
+ccflags-y += -Wno-incompatible-pointer-types
+ccflags-y += -Wno-unused-variable
+
```
these are good warnings to enable, but unrelated to the feature you are implementing. Perhaps they should be in a separate commit ?

mmakassikis commented on mgmt/user_config.h:
```diff
@@ -27,6 +27,13 @@ static inline bool user_guest(struct ksmbd_user *user)
+{
+	if (user->name[0] == '\0')
+		return 1;
+	return 0;
```
this can be shortened to:
> return user->name[0] == '\0';

and the return type should be bool, no ?

mmakassikis commented on mgmt/user_session.c:
```diff
@@ -27,6 +30,194 @@ struct ksmbd_session_rpc {
+	return session->user->name;
+}
+
+static int show_proc_session(struct seq_file *m, void *v)
```
Isn't this function racy ?
the commented `get_session()` / `put_session()` hint that some locking to protect `struct ksmbd_session` from being freed.
Perhaps you meant `ksmbd_user_session_get()` and `ksmbd_user_session_put()` ?

mmakassikis commented on mgmt/user_session.c:
```diff
@@ -27,6 +30,194 @@ struct ksmbd_session_rpc {
+	seq_printf(m, "#%-40s %-15s %-10s %-10s\n",
+		 "<client>", "<user>", "<sess_id>", "<state>");
+
+	down_read(&sessions_table_lock);
```
I don't know if this is enough. Code in mgmt/user_session.c that locks `sessions_table_lock` also locks the `conn->session_lock`.

mmakassikis commented on mgmt/user_session.c:
```diff
@@ -27,6 +30,194 @@ struct ksmbd_session_rpc {
+
+	down_read(&sessions_table_lock);
+	hash_for_each(sessions_table, i, session, hlist) {
+		// get_session(session);
```
Is that a leftover, or do you need to use ksmbd_user_session_get() and ksmbd_user_session_put() ?

mmakassikis commented on proc.c:
```diff
@@ -0,0 +1,149 @@
+#define SMB2_CHANGE_NOTIFY		0x0000000F
+#define SMB2_QUERY_INFO		0x00000010
+#define SMB2_SET_INFO		0x00000011
+#define SMB2_OPLOCK_BREAK		0x00000012
```
Aren't these the same definitions as the ones in fs/smb/common/smb2pdu.h ?

mmakassikis commented on proc.c:
```diff
@@ -0,0 +1,149 @@
+	seq_printf(m, "max protocol: %hd\n", server_conf.max_protocol);
+	seq_printf(m, "flags: 0x%08x\n", server_conf.flags);
+	seq_printf(m, "share_fake_fscaps: 0x%08x\n",
+		   server_conf.share_fake_fscaps);
```
nitpick (so feel free to ignore it): the values above are configuration rather than stats. It's still useful to be able to consult at runtime.

Is it possible to have human-readable values for min_protocol/max_protocol ?

mmakassikis commented on smb2pdu.c:
```diff
@@ -1501,7 +1508,7 @@ static int ntlm_authenticate(struct ksmbd_work *work,
 		 * on reauthetication.
 		 */
-		if (conn->binding == false && ksmbd_anonymous_user(user)) {
+		if (conn->binding == false && user_anonymous(user)) {
```
I hadn't noticed user_anonymous() is ksmbd_anonymous_user(). I guess it's more coherent with user_guest(), but shouldn't that be a separate commit ?

mmakassikis commented on stats.h:
```diff
@@ -0,0 +1,75 @@
+#ifndef __KSMBD_STATS_H__
+#define __KSMBD_STATS_H__
+
+#define KSMBD_COUNTER_MAX_REQS	19
```
Can you add an explanation as to where the magic value comes from ? (even better: is it possible to get rid of it ?)

mmakassikis commented on stats.h:
```diff
@@ -0,0 +1,75 @@
+#else
+struct ksmbd_counters {
+	int counters[0];
+};
```
is it necessary to declare ksmbd_counters in the !PROCFS case ?

mmakassikis commented on transport_tcp.c:
```diff
@@ -247,8 +248,16 @@ static int ksmbd_tcp_new_connection(struct socket *client_sk)
+
 	handler = kthread_run(ksmbd_conn_handler_loop,
-			      KSMBD_TRANS(t)->conn,
+			      conn,
```
this looks like code from a different commit

mmakassikis commented on vfs_cache.c:
```diff
@@ -37,6 +38,106 @@ static struct ksmbd_file_table global_ft;
+#define SMB2_LEASE_NONE_LE			cpu_to_le32(0x00)
+#define SMB2_LEASE_READ_CACHING_LE		cpu_to_le32(0x01)
+#define SMB2_LEASE_HANDLE_CACHING_LE		cpu_to_le32(0x02)
+#define SMB2_LEASE_WRITE_CACHING_LE		cpu_to_le32(0x04)
```
use the values from fs/smb/common/smb2pdu.h

mmakassikis reviewed (commented):
A few comments from reading the code. I haven't compiled it or checked it for correctness.

vedansh-bhardwaj commented:
Hi @namjaejeon we fixed the warnings and errors except this below mentioned warning.
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#568:
new file mode 100644

Please review.

Thanks

namjaejeon commented:
@vedansh-bhardwaj Okay, I will check it. Thanks!

namjaejeon commented on mgmt/user_session.c:
```diff
@@ -538,3 +735,9 @@ void ksmbd_release_tree_conn_id(struct ksmbd_session *sess, int id)
 		ksmbd_release_id(&sess->tree_conn_ida, id);
 }
+
+int ksmbd_sessions_init(void)
```
we don't need this wrapper function. please open code and directly use create_proc_sessions().

vedansh-bhardwaj replied:
Fixed

namjaejeon commented on connection.h:
```diff
@@ -114,6 +115,9 @@ struct ksmbd_conn {
 	__le16				signing_algorithm;
 	bool				binding;
+
+	char				client_name[INET6_ADDRSTRLEN + 1];
```
Should it consume 49bytes per connection ? can we use inet6_addr or inet_addr ?

namjaejeon commented on mgmt/user_session.c:
```diff
@@ -27,6 +30,195 @@ struct ksmbd_session_rpc {
+	else if (session->state == SMB2_SESSION_EXPIRED)
+		return "expired";
+	else
+		return "";
```
How about use switch condition instead of if ?

namjaejeon commented on mgmt/user_session.c:
```diff
@@ -27,6 +30,195 @@ struct ksmbd_session_rpc {
+	}
+
+	i = 0;
+	xa_for_each(&sess->ksmbd_chann_list, id, chan) {
```
we can remove {} if loop is single line.

vedansh-bhardwaj replied:
Fixed

namjaejeon commented on proc.c:
```diff
@@ -0,0 +1,131 @@
+
+struct proc_dir_entry *ksmbd_proc_create(const char *name,
+					 int (*show)(struct seq_file *m, void *v),
+			     void *v)
```
I prefer two tab and align it

namjaejeon commented on smb2pdu.c:
```diff
@@ -231,6 +232,11 @@ u16 get_smb2_cmd_val(struct ksmbd_work *work)
 	return le16_to_cpu(rcv_hdr->Command);
 }
 
+void smb2_inc_reqs(unsigned int cmd)
```
ditto, we can remove wrapper function and open code.

namjaejeon commented on vfs_cache.c:
```diff
@@ -37,6 +39,100 @@ static struct ksmbd_file_table global_ft;
+
+static const struct ksmbd_const_name ksmbd_lease_const_names[] = {
+	{le32_to_cpu(SMB2_LEASE_NONE_LE),
+		"LEASE_NONE"},
```
Is there any reason why you put "LEASE_NONE" in next line ? the line is over 80 ?

vedansh-bhardwaj commented:
Hi @namjaejeon 
I will do the needful changes and update you.
Thanks

namjaejeon commented:
@vedansh-bhardwaj Okay, Looks good to me, BTW, there are so many signed-off-by and authors.

Signed-off-by: v.bhardwaj <v.bhardwaj@samsung.com>
Signed-off-by: bahubali.bg <bahubali.bg@samsung.com>
Signed-off-by: m.sataraddi <m.sataraddi@samsung.com>
Signed-off-by: c.singh <c.singh@samsung.com>
Signed-off-by: constant.lee <constant.lee@samsung.com>
Signed-off-by: moons49.kim <moons49.kim@samsung.com>


+ *   Author(s):v.bhardwaj <v.bhardwaj@samsung.com>
+ *   Author(s):bahubali.bg <bahubali.bg@samsung.com>
+ *   Author(s):m.sataraddi <m.sataraddi@samsung.com>
+ *   Author(s):c.singh <c.singh@samsung.com>
+ *   Author(s):constant.lee <constant.lee@samsung.com>
+ *   Author(s):moons49.kim <moons49.kim@samsung.com>

I don't think all of these people were involved. Please make signed-off-by up to 2 people and author up to 1 person.
Thanks.

vedansh-bhardwaj commented:
Hi @namjaejeon we will update.
Thanks

vedansh-bhardwaj commented:
Hi @namjaejeon 
updated the signed-off and authors details. Please review.
Thanks

namjaejeon commented:
@vedansh-bhardwaj I have updated your patch like this (https://github.com/namjaejeon/ksmbd/commit/18abb79ea54ea482eb5c2ee7b62e47a4b859e6ec).
I changed %pI6 and %pl4 to %pl6c and %pl4c. It will not show hex ip address. Is it okay for your target ?

vedansh-bhardwaj commented:
Hi @namjaejeon 
It is ok for us.
Thanks

vedansh-bhardwaj commented:
Hi @namjaejeon 
Happy new year

Any update on these proc related changes.

thanks

namjaejeon commented:
@vedansh-bhardwaj The proc patches are in next branch. It will be applied to master when linux 6.20-rc1 merge window open.

vedansh-bhardwaj commented:
Hi @namjaejeon 
Thanks for update.
